x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -43,698 +39,534 @@
-            );
 
-            // TODO The implementation of JWT authentication will be completed in a later PR
-            // At this point, this is the right realm to do validation. Trigger AuthenticationResult.unsuccessful() if any problems found.
-
-            // Cache lookup
-            // 1. If present in cache, AuthenticationResult.successful().
-
-            // Client Authorization
-            switch (this.clientAuthorizationType) {
-                case JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET:
-                    if (Strings.hasText(clientAuthorizationSharedSecretString) == false) {
-                        final String msg = String.format(
-                            Locale.ROOT,
-                            "Realm [%s] client authentication [%s] failed for tokenPrincipal [%s] because request header is missing.",
-                            super.name(),
-                            this.clientAuthorizationType,
-                            tokenPrincipal
-                        );
-                        LOGGER.debug(msg);
-                        listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
-                        return;
-                    } else if (this.clientAuthorizationSharedSecret.toString().equals(clientAuthorizationSharedSecretString) == false) {
-                        final String msg = String.format(
-                            Locale.ROOT,
-                            "Realm [%s] client authentication [%s] failed for tokenPrincipal [%s] because request header did not match.",
-                            super.name(),
-                            this.clientAuthorizationType,
-                            tokenPrincipal
-                        );
-                        LOGGER.debug(msg);
-                        listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
-                        return;
+            // Validate JWT: Only for JWT Cache(Miss) or Off. Not for Hit(SigOnly). Hit(Pass|Fail) never falls through to here.
+            final String jwtAlg = jwt.getHeader().getAlgorithm().getName();
+            if (jwtCacheMissOrOff) {
+                try {
+                    final boolean isJwtAlgHmac = JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_HMAC.contains(jwtAlg);
+                    final List<String> algs = isJwtAlgHmac ? this.algorithmsHmac : this.algorithmsPkc;
+                    final List<JWK> jwks = isJwtAlgHmac ? this.jwksHmac : this.jwksPkc;
+                    JwtValidateUtil.validate(jwt, this.allowedIssuer, this.allowedAudiences, this.allowedClockSkew.seconds(), algs, jwks);
+                    LOGGER.trace("Realm [" + super.name() + "] JWT validation succeeded for token=[" + tokenPrincipal + "].");
+                    if (this.jwtValidationCache != null) {
+                        this.jwtValidationCache.put(jwtValidationCacheKey, Optional.empty()); // JWT Cache Hit(SigOnly)
```

Comment:
Why put this Optional.empty() in the realm here and _then_ update it with a success/failure after the role mapping ? Why not wait until we go to role mapping and then just add a success / failure to the cache ? Do we gain anything by this intermediate result in the cache ? It feels that it would be a big enough simplification to the code